providers/common/bio_prov.c

```diff
@@ -94,3 +109,110 @@ int ossl_prov_bio_printf(BIO *bio, const char *format, ...)
     return ret;
 }
 
+#ifndef FIPS_MODULE
+
+/* No direct BIO support in the FIPS module */
+
+static int bio_core_read_ex(BIO *bio, char *data, size_t data_len,
+                            size_t *bytes_read)
+{
+    return ossl_prov_bio_read_ex(BIO_get_data(bio), data, data_len, bytes_read);
+}
+
+static int bio_core_write_ex(BIO *bio, const char *data, size_t data_len,
+                             size_t *written)
+{
+    return ossl_prov_bio_write_ex(BIO_get_data(bio), data, data_len, written);
+}
+
+static long bio_core_ctrl(BIO *bio, int cmd, long num, void *ptr)
+{
+    /* We don't support this */
+    assert(0);
```

Comment:
This may be foolish.  What if, for example, someone wants to `BIO_eof(bio)`?  It would be perfectly reasonable to expect that to work properly...

Considering that the BIO ctrl commands have been stable for eons, I don't expect this to become a can of worms.

Reply:
This is not public API. This is about providing a BIO for our own providers to use. If we need support for additional ctrls then we can add it at that point. Probably I should have called the BIO creation function `bio_new_from_core_bio` i.e. with a lowercase `bio` to emphasise that point.

Reply:
The trouble is that *something* inside the provider, or in libcrypto that the provider calls, might assume a normal BIO.

Oh well, I guess we'll get there some day...